In [2]:
from tabl import *

# THE BASICS

In [3]:
stirling_set(5, 3)

25

In [4]:
stirling_set(5)

[0, 1, 15, 25, 10, 1]

In [5]:
stirling_set.tab(9)

[[1],
 [0, 1],
 [0, 1, 1],
 [0, 1, 3, 1],
 [0, 1, 7, 6, 1],
 [0, 1, 15, 25, 10, 1],
 [0, 1, 31, 90, 65, 15, 1],
 [0, 1, 63, 301, 350, 140, 21, 1],
 [0, 1, 127, 966, 1701, 1050, 266, 28, 1]]

In [6]:
stirling_set.rev(9)

[[1],
 [1, 0],
 [1, 1, 0],
 [1, 3, 1, 0],
 [1, 6, 7, 1, 0],
 [1, 10, 25, 15, 1, 0],
 [1, 15, 65, 90, 31, 1, 0],
 [1, 21, 140, 350, 301, 63, 1, 0],
 [1, 28, 266, 1050, 1701, 966, 127, 1, 0]]

In [7]:
stirling_set.inv(9)

[[1],
 [0, 1],
 [0, -1, 1],
 [0, 2, -3, 1],
 [0, -6, 11, -6, 1],
 [0, 24, -50, 35, -10, 1],
 [0, -120, 274, -225, 85, -15, 1],
 [0, 720, -1764, 1624, -735, 175, -21, 1],
 [0, -5040, 13068, -13132, 6769, -1960, 322, -28, 1]]

In [8]:
stirling_set.revinv(9)

[[1],
 [1, 0],
 [1, -1, 0],
 [1, -3, 2, 0],
 [1, -6, 11, -6, 0],
 [1, -10, 35, -50, 24, 0],
 [1, -15, 85, -225, 274, -120, 0],
 [1, -21, 175, -735, 1624, -1764, 720, 0],
 [1, -28, 322, -1960, 6769, -13132, 13068, -5040, 0]]

'stirling_set.invrev', i.e. the inverse of the reversed stirling_set triangle does not exist because stirling_set.rev is not invertible. In such cases simply the empty triangle [] is returned (and no error thrown).

In [9]:
stirling_set.invrev(9)

[]

Conversely, a triangle which has not a 'revinv' but an 'invrev' form is the Delannoy triangle. Of course, all this depends on how the triangle was defined in the first place. And indeed some call the reversed Delannoy triangle the Delannoy triangle. 

In [10]:
delannoy.invrev(9)

[[1],
 [-1, 1],
 [2, -3, 1],
 [-6, 10, -5, 1],
 [22, -38, 22, -7, 1],
 [-90, 158, -98, 38, -9, 1],
 [394, -698, 450, -194, 58, -11, 1],
 [-1806, 3218, -2126, 978, -334, 82, -13, 1],
 [8558, -15310, 10286, -4942, 1838, -526, 110, -15, 1]]

'stirling_set.mat' converts the triangle to a square matrix by filling T(n, k) for k > n with zeros.

In [11]:
stirling_set.mat(9)

[[1, 0, 0, 0, 0, 0, 0, 0, 0],
 [0, 1, 0, 0, 0, 0, 0, 0, 0],
 [0, 1, 1, 0, 0, 0, 0, 0, 0],
 [0, 1, 3, 1, 0, 0, 0, 0, 0],
 [0, 1, 7, 6, 1, 0, 0, 0, 0],
 [0, 1, 15, 25, 10, 1, 0, 0, 0],
 [0, 1, 31, 90, 65, 15, 1, 0, 0],
 [0, 1, 63, 301, 350, 140, 21, 1, 0],
 [0, 1, 127, 966, 1701, 1050, 266, 28, 1]]

The identifier of the function. 

In [12]:
stirling_set.id

'STIRLING2SET'

# TABLES

In [13]:
t = stirling_set.tab(9)
PrintFlat(t)

[1, 0, 1, 0, 1, 1, 0, 1, 3, 1, 0, 1, 7, 6, 1, 0, 1, 15, 25, 10, 1, 0, 1, 31, 90, 65, 15, 1, 0, 1, 63, 301, 350, 140, 21, 1, 0, 1, 127, 966, 1701, 1050, 266, 28, 1]


In [14]:
t = stirling_set.tab(9)
PrintTabl(t)

[[1], [0, 1], [0, 1, 1], [0, 1, 3, 1], [0, 1, 7, 6, 1], [0, 1, 15, 25, 10, 1], [0, 1, 31, 90, 65, 15, 1], [0, 1, 63, 301, 350, 140, 21, 1], [0, 1, 127, 966, 1701, 1050, 266, 28, 1]]


In [15]:
t = stirling_set.tab(4)
PrintTerms(t)

0 [0, 0] 1
1 [1, 0] 0
2 [1, 1] 1
3 [2, 0] 0
4 [2, 1] 1
5 [2, 2] 1
6 [3, 0] 0
7 [3, 1] 1
8 [3, 2] 3
9 [3, 3] 1


In [ ]:
t = stirling_set.tab(9)
PrintRows(t)

| trow  |  list  |
| :---  |  :---  |
| trow0 | [1] |
| trow1 | [0, 1] |
| trow2 | [0, 1, 1] |
| trow3 | [0, 1, 3, 1] |
| trow4 | [0, 1, 7, 6, 1] |
| trow5 | [0, 1, 15, 25, 10, 1] |
| trow6 | [0, 1, 31, 90, 65, 15, 1] |
| trow7 | [0, 1, 63, 301, 350, 140, 21, 1] |
| trow8 | [0, 1, 127, 966, 1701, 1050, 266, 28, 1] |

In [ ]:
PrintRowArray(stirling_set, 6, 7)

| rdiag  |   seq  |
| :---   |  :---  |
| rdiag0 | [1, 1, 1, 1, 1, 1, 1]|
| rdiag1 | [0, 1, 3, 6, 10, 15, 21]|
| rdiag2 | [0, 1, 7, 25, 65, 140, 266]|
| rdiag3 | [0, 1, 15, 90, 350, 1050, 2646]|
| rdiag4 | [0, 1, 31, 301, 1701, 6951, 22827]|
| rdiag5 | [0, 1, 63, 966, 7770, 42525, 179487]|
| rdiag6 | [0, 1, 127, 3025, 34105, 246730, 1323652]|

In [ ]:
PrintColArray(stirling_set, 6, 7) 

| cdiag  |   seq  |
| :---   |  :---  |
| cdiag0 | [1, 0, 0, 0, 0, 0] |
| cdiag1 | [1, 1, 1, 1, 1, 1] |
| cdiag2 | [1, 3, 7, 15, 31, 63] |
| cdiag3 | [1, 6, 25, 90, 301, 966] |
| cdiag4 | [1, 10, 65, 350, 1701, 7770] |
| cdiag5 | [1, 15, 140, 1050, 6951, 42525] |
| cdiag6 | [1, 21, 266, 2646, 22827, 179487] |

In [ ]:
PrintRowPolyArray(stirling_set, 6, 7)

| rpdiag  |   seq  |
| :---    |  :---  |
| rpdiag0 | [1, 1, 1, 1, 1, 1, 1] |
| rpdiag1 | [0, 1, 2, 3, 4, 5, 6] |
| rpdiag2 | [0, 2, 6, 12, 20, 30, 42] |
| rpdiag3 | [0, 5, 22, 57, 116, 205, 330] |
| rpdiag4 | [0, 15, 94, 309, 756, 1555, 2850] |
| rpdiag5 | [0, 52, 454, 1866, 5428, 12880, 26682] |

In [ ]:
PrintColPolyArray(stirling_set, 6, 7)

| cpdiag  |   seq  |
| :---    |  :---  |
| cpdiag0 | [1, 0, 0, 0, 0, 0, 0] |
| cpdiag1 | [1, 1, 2, 5, 15, 52, 203] |
| cpdiag2 | [1, 2, 6, 22, 94, 454, 2430] |
| cpdiag3 | [1, 3, 12, 57, 309, 1866, 12351] |
| cpdiag4 | [1, 4, 20, 116, 756, 5428, 42356] |
| cpdiag5 | [1, 5, 30, 205, 1555, 12880, 115155] |

In [ ]:
PrintViews(abel)


ABELPOLYNOMS Triangle view

| trow  |  list  |
| :---  |  :---  |
| trow0 | [1] |
| trow1 | [0, 1] |
| trow2 | [0, 2, 1] |
| trow3 | [0, 9, 6, 1] |
| trow4 | [0, 64, 48, 12, 1] |
| trow5 | [0, 625, 500, 150, 20, 1] |
| trow6 | [0, 7776, 6480, 2160, 360, 30, 1] |

ABELPOLYNOMS Flattened seqs

| flat      |   seq  |
| :---      |  :---  |
| tabl     | [1, 0, 1, 0, 2, 1, 0, 9, 6, 1, 0, 64, 48, 12, 1, 0, 625, 500, 150, 20, 1, 0, 7776, 6480, 2160, 360, 30, 1] |
| rev      | [1, 1, 0, 1, 2, 0, 1, 6, 9, 0, 1, 12, 48, 64, 0, 1, 20, 150, 500, 625, 0, 1, 30, 360, 2160, 6480, 7776, 0] |
| cum      | [1, 0, 1, 0, 2, 3, 0, 9, 15, 16, 0, 64, 112, 124, 125, 0, 625, 1125, 1275, 1295, 1296, 0, 7776, 14256, 16416, 16776, 16806, 16807] |
| revcum   | [1, 1, 0, 3, 2, 0, 16, 15, 9, 0, 125, 124, 112, 64, 0, 1296, 1295, 1275, 1125, 625, 0, 16807, 16806, 16776, 16416, 14256, 7776, 0] |
| cumrev   | [1, 1, 1, 1, 3, 3, 1, 7, 16, 16, 1, 13, 61, 125, 125, 1, 21, 171, 671, 1296, 1296, 1, 31, 391, 2551, 9031, 16807, 16807] |
| diag     | [1, 0, 0, 1, 0, 2, 0, 9, 1, 0, 64, 6] |

ABELPOLYNOMS Row sums

| sum       |   seq  |
| :---      |  :---  |
| sum       | [1, 1, 3, 16, 125, 1296, 16807] |
| evensum   | [1, 0, 1, 6, 49, 520, 6841] |
| oddsum    | [0, 1, 2, 10, 76, 776, 9966] |
| altsum    | [1, -1, -1, -4, -27, -256, -3125] |
| diagsum   | [1, 0, 1, 2, 10, 70] |
| cumsum    | [1, 1, 5, 40, 425, 5616, 88837] |
| revcumsum | [1, 2, 7, 40, 325, 3456, 45619] |

ABELPOLYNOMS Diagonals as rows

| rdiag  |   seq  |
| :---   |  :---  |
| rdiag0 | [1, 1, 1, 1, 1, 1, 1]|
| rdiag1 | [0, 2, 6, 12, 20, 30, 42]|
| rdiag2 | [0, 9, 48, 150, 360, 735, 1344]|
| rdiag3 | [0, 64, 500, 2160, 6860, 17920, 40824]|
| rdiag4 | [0, 625, 6480, 36015, 143360, 459270, 1260000]|
| rdiag5 | [0, 7776, 100842, 688128, 3306744, 12600000, 40584852]|
| rdiag6 | [0, 117649, 1835008, 14880348, 84000000, 372027810, 1379524608]|

ABELPOLYNOMS Diagonals as columns

| cdiag  |   seq  |
| :---   |  :---  |
| cdiag0 | [1, 0, 0, 0, 0, 0, 0] |
| cdiag1 | [1, 2, 9, 64, 625, 7776, 117649] |
| cdiag2 | [1, 6, 48, 500, 6480, 100842, 1835008] |
| cdiag3 | [1, 12, 150, 2160, 36015, 688128, 14880348] |
| cdiag4 | [1, 20, 360, 6860, 143360, 3306744, 84000000] |
| cdiag5 | [1, 30, 735, 17920, 459270, 12600000, 372027810] |
| cdiag6 | [1, 42, 1344, 40824, 1260000, 40584852, 1379524608] |

ABELPOLYNOMS Polynomial values as rows

| rpdiag  |   seq  |
| :---    |  :---  |
| rpdiag0 | [1, 1, 1, 1, 1, 1, 1] |
| rpdiag1 | [0, 1, 2, 3, 4, 5, 6] |
| rpdiag2 | [0, 3, 8, 15, 24, 35, 48] |
| rpdiag3 | [0, 16, 50, 108, 196, 320, 486] |
| rpdiag4 | [0, 125, 432, 1029, 2048, 3645, 6000] |
| rpdiag5 | [0, 1296, 4802, 12288, 26244, 50000, 87846] |
| rpdiag6 | [0, 16807, 65536, 177147, 400000, 805255, 1492992] |

ABELPOLYNOMS Polynomial values as columns

| cpdiag  |   seq  |
| :---    |  :---  |
| cpdiag0 | [1, 0, 0, 0, 0, 0, 0] |
| cpdiag1 | [1, 1, 3, 16, 125, 1296, 16807] |
| cpdiag2 | [1, 2, 8, 50, 432, 4802, 65536] |
| cpdiag3 | [1, 3, 15, 108, 1029, 12288, 177147] |
| cpdiag4 | [1, 4, 24, 196, 2048, 26244, 400000] |
| cpdiag5 | [1, 5, 35, 320, 3645, 50000, 805255] |
| cpdiag6 | [1, 6, 48, 486, 6000, 87846, 1492992] |
